### PR TITLE
Modernize the thunderbolt plugin

### PR DESCRIPTION
**Pull request by superm1**
WIP, but I'm hoping this will be ready for 1.4.0.  If the controller GUID thing looks good, I probably want that in 1_3_X too.

There are various TODO left, so any thoughts on those welcome.  The biggest TODO left in my mind is force power handling.

Any other things that should be done to modernize also welcome.

Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/hughsie/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by superm1** [comment]
CC @YehezkelShB

**Comment by superm1** [comment]
(Oh and yes I know self tests are busted, need to figure that out still too..)

**Comment by hughsie** [comment]
> Modernize the thunderbolt plugin

This has been on my "really ought to fix" list for a long while, thanks for finally sorting it. I think you and @gicmo understand this stuff much better than me anyway.

**Comment by superm1** [comment]
Sure.

Do you think you might be able to help with conversion to use fufirmware? I think you understand that part better at least.

**Comment by hughsie** [comment]
> Do you think you might be able to help with conversion to use fufirmware

I think I understand what's going on now, I'll work in the firmware parser on Monday.

**Comment by hughsie** [comment]
Current status:

    sudo ./src/fwupdtool firmware-parse TR_*0x09E1.bin thunderbolt
    FuThunderboltFirmware:
    IsHost:                  true
    IsNative:                true
    DeviceId:                0x15e7
    VendorId:                0xd4
    ModelId:                 0x9e1
    Generation:              0x3
    Ports:                   0x1
    HasPd:                   true
    Section0:                0x4000
    Section1:                0x4210
    Section2:                0x4610
    Section3:                0x271f0
      FuFirmwareImage:
      Data:                  0x66000

File: plugins/thunderbolt/fu-thunderbolt-firmware.c [hunk 1]
```diff
@@ -734,54 +749,59 @@ fu_thunderbolt_image_validate (GBytes  *controller_fw,
...
 	 * good enough validation.
 	 */
-	if (hw_info->id == 0)
-		return UNKNOWN_DEVICE;
+	if (hw_info->id == 0) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_NOT_SUPPORTED,
+				     "no supported");
+		return FALSE;
```
Comment (YehezkelShB) [inline plugins/thunderbolt/fu-thunderbolt-firmware.c]: So we decided we don't want to support newer devices if the host support hasn't added yet?
At least let's remove the comment that says we do want to support it.

**Comment by superm1** [comment]
> Just another comment (besides the one I spotted in the code): I see you dropped native and safe-mode. This means we drop support for old (i.e. just couple of years old) HW? Is this considered normal for fwupd?

Can you point out in the code where you notice this should be?  That wasn't the intention.  It was just to drop it from the reporting metadata to LVFS.

**Comment by superm1** [comment]
> BTW, the last things I knew, the plan was to continue producing add-in cards, which means safe-mode keeps being relevant. I'm sure your knowledge is much more up-to-date, so I assume there are no new add-in cards since the integration to the CPU (and no add-in cards planned for USB4 too)?

I don't believe I can publicly talk about unreleased future stuff.  But I agree we should keep supporting all features that were previously present no matter what is happening for the future.

**Comment by YehezkelShB** [comment]
> > Just another comment (besides the one I spotted in the code): I see you dropped native and safe-mode. This means we drop support for old (i.e. just couple of years old) HW? Is this considered normal for fwupd?
> 
> Can you point out in the code where you notice this should be? That wasn't the intention. It was just to drop it from the reporting metadata to LVFS.

Sorry, I misunderstood your commit message.
If by "reporting metadata to LVFS" you mean the GUID generation (to select the correct FW image file) still takes "native" flag into consideration, it's just the statistic data collection that drops it, I'm totally fine with it.

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 2]
```diff
@@ -0,0 +1,645 @@
...
+	}
+}
+
+static void
+fu_thunderbolt_device_to_string (FuDevice *device, guint idt, GString *str)
+{
+	FuThunderboltDevice *self = FU_THUNDERBOLT_DEVICE (device);
+	fu_common_string_append_kb (str, idt, "Host Controller", self->host);
+	fu_common_string_append_kb (str, idt, "Safe Mode", self->safe_mode);
+}
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c]: Since the FW is being parsed at daemon startup now we may as well also add the following:
* "Is Native"
* Generation

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 3]
```diff
@@ -0,0 +1,645 @@
...
+		return NULL;
+
+	/* get current NVMEM */
+	nvmem = fu_thunderbolt_device_find_nvmem (self, TRUE, error);
+	if (nvmem == NULL)
+		return NULL;
+	controller_fw = g_file_load_bytes (nvmem, NULL, NULL, error);
+	if (!fu_firmware_parse (FU_FIRMWARE (firmware_old), controller_fw, flags, error))
+		return NULL;
+	if (fu_thunderbolt_firmware_is_host (firmware) !=
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c]: Since there is a way to tell if it's host by reading the FW image itself now - maybe we should use that at startup rather than the examining of the udev parent?  This seems safer.

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 4]
```diff
@@ -0,0 +1,645 @@
...
+	    fu_thunderbolt_firmware_get_device_id (firmware_old)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INVALID_FILE,
+			     "incorrect device type, got 0x%04x, expected 0x%04x",
+			     fu_thunderbolt_firmware_get_device_id (firmware),
+			     fu_thunderbolt_firmware_get_device_id (firmware_old));
+		return NULL;
+	}
+	if ((flags & FWUPD_INSTALL_FLAG_FORCE) == 0) {
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c]: I would say we may as well just move this to the top of the function and let force override anything, not just the model ID check?

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 5]
```diff
@@ -0,0 +1,645 @@
...
+			     FWUPD_ERROR_INTERNAL,
+			     "update failed (status %" G_GINT64_MODIFIER "x)", status);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_thunderbolt_device_write_data (FuThunderboltDevice	*self,
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c]: This function I wonder if we should replace the code with FuChunks instead

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 6]
```diff
@@ -0,0 +1,645 @@
...
+		if (domain_id != NULL)
+			fu_device_add_instance_id (device, domain_id);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_plugin_thunderbolt_trigger_update (FuThunderboltDevice	*self,
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c]: Something I'm considering here is if we should instead kick this into a delayed activation (or maybe make that behavior optional).

If we did that then potentially then thunderbolt devices could set a flag that they are usable during an update and once we reboot/shutdown apply the update.

The downside is unplugging a Thunderbolt or USB4 device without rebooting will cause firmware to not be updated.  So it is definitely a trade off.

File: plugins/thunderbolt/fu-thunderbolt-firmware.c [hunk 7]
```diff
@@ -0,0 +1,517 @@
...
+							  &pd_pointer,
+							  error)) {
+			g_prefix_error (error, "failed to read pd-pointer: ");
+			return FALSE;
+		}
+		self->has_pd = fu_thunderbolt_firmware_valid_pd_pointer (pd_pointer);
+	}
+
+	if (!self->is_host) {
+		/* TODO: is this to be multiplied by a multiplier, e.g. 0x10000? */
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-firmware.c]: @YehezkelShB do you know?

File: plugins/thunderbolt/fu-thunderbolt-firmware.c [hunk 8]
```diff
@@ -0,0 +1,517 @@
...
+	return TRUE;
+}
+
+/* assumes sections[_SECTION_DIGITAL].offset is already set */
+static gboolean
+fu_thunderbolt_firmware_read_sections (FuThunderboltFirmware *self, GError **error)
+{
+	guint32 offset;
+
+	if (self->gen >= 3 || self->gen == 0) {
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-firmware.c]: I don't know that we should check > 3 unless this can be checked on a real Intel USB4 payload.

File: plugins/thunderbolt/fu-thunderbolt-device.c (hunk 5 above)
Comment (hughsie) [reply plugins/thunderbolt/fu-thunderbolt-device.c]: I don't think we can, it's not writing in blocks of a set size, it's just writing until write() returns and then advancing a counter with the number of bytes we wrote.

File: plugins/thunderbolt/fu-thunderbolt-device.c (hunk 4 above)
Comment (hughsie) [reply plugins/thunderbolt/fu-thunderbolt-device.c]: do we ever want to allow installing a host firmware on a device? I'd argue we just let the user blow their feet off... Maybe VID and PID makes sense to ignore tho?

File: plugins/thunderbolt/fu-thunderbolt-device.c (hunk 6 above)
Comment (hughsie) [reply plugins/thunderbolt/fu-thunderbolt-device.c]: I think if we can apply it live we can -- rebooting is a real pain the neck for most users.

File: plugins/thunderbolt/fu-thunderbolt-firmware.c (hunk 8 above)
Comment (hughsie) [reply plugins/thunderbolt/fu-thunderbolt-firmware.c]: ¯\_(ツ)_/¯ I just copied it from the old source code. I don't really understand what it's doing...

File: plugins/thunderbolt/fu-thunderbolt-device.c (hunk 4 above)
Comment (superm1) [reply plugins/thunderbolt/fu-thunderbolt-device.c]: The only circumstance that this was needed I know of was a mid-development model ID string change.  Maybe you're right - we shouldn't let people shoot themselves in the foot until we know another real need for it.

File: plugins/thunderbolt/fu-thunderbolt-firmware.c (hunk 8 above)
Comment (superm1) [reply plugins/thunderbolt/fu-thunderbolt-firmware.c]: oh, okay lets leave it as is for now then and revisit when we see some usb4 stuff in the wild.

File: plugins/thunderbolt/fu-thunderbolt-device.c (hunk 3 above)
Comment (hughsie) [reply plugins/thunderbolt/fu-thunderbolt-device.c]: Is we want to find out if it's host we have to read a *lot* more of the image. 

    failed to read is-host: location is outside of the given image: attempted to read 0x01 bytes at offset 0x7f014b from buffer of 0x80

I've pushed a commit that at least makes use of the native boolean.

Comment (superm1) [reply plugins/thunderbolt/fu-thunderbolt-device.c]: Sounds good, thanks.

File: plugins/thunderbolt/fu-thunderbolt-firmware.c (hunk 7 above)
Comment (YehezkelShB) [reply plugins/thunderbolt/fu-thunderbolt-firmware.c]: There is a reason why it was just one another location in the comparisons done for the images. This is just an "enum"-like value. There is no point with trying to treat it as a "real" size.

File: plugins/thunderbolt/fu-thunderbolt-firmware.c [hunk 9]
```diff
@@ -0,0 +1,521 @@
...
+							  0x10C,
+							  &pd_pointer,
+							  error)) {
+			g_prefix_error (error, "failed to read pd-pointer: ");
+			return FALSE;
+		}
+		self->has_pd = fu_thunderbolt_firmware_valid_pd_pointer (pd_pointer);
+	}
+
+	if (!self->is_host) {
```
Comment (YehezkelShB) [inline plugins/thunderbolt/fu-thunderbolt-firmware.c]: In the original code, this was checked and compared only for host, and not for all of them (just from AR). This wasn't a mistake. As I said, it's hard to follow all the changes done here in this massive code move, but any change that is more than pure refactoring, e.g. checking more things than was checked before or less things, isn't guaranteed to work properly.

**Comment by YehezkelShB** [comment]
Seems like you decided to get rid of all the FW comparisons made for host FW, I can't find them in the new code. What am I missing?

File: plugins/thunderbolt/fu-thunderbolt-firmware.c (hunk 8 above)
Comment (YehezkelShB) [reply plugins/thunderbolt/fu-thunderbolt-firmware.c]: I think the assumption was that there will be new gen after 3, but these locations were meant to be kept the same. I don't see any harm in keeping it like this. New gen will not be supported automatically anyway, we have to add it to the controller ID list first.

**Comment by hughsie** [comment]
> Seems like you decided to get rid of all the FW comparisons made for host FW

This was deliberate. The FuFirmware validation in all the other plugins is really just an end-user sanity check, rather than a system integrator check. In all honesty checking all the other fields is way overkill for what we're doing, and also means we have to update a big table of data for every generation change.

File: plugins/thunderbolt/fu-thunderbolt-firmware.c (hunk 9 above)
Comment (hughsie) [reply plugins/thunderbolt/fu-thunderbolt-firmware.c:487]: Understood. @superm1 do you think we should just remove the flash_size completely? It doesn't seem super-useful.

**Comment by YehezkelShB** [comment]
> > Seems like you decided to get rid of all the FW comparisons made for host FW
> 
> This was deliberate. The FuFirmware validation in all the other plugins is really just an end-user sanity check, rather than a system integrator check. In all honesty checking all the other fields is way overkill for what we're doing, and also means we have to update a big table of data for every generation change.

I hope you talked to FW team to hear of they are OK with it. One of the real issues is that not all host FW include DROM. Especially, pre-AR FWs didn't have it at all. So you can't trust the vendor and model IDs to save you from flashing incompatible FW. I recommend to keep all those comparisons, even if they don't look like the style you'd like for "modern" plugin.

**Comment by hughsie** [comment]
> So you can't trust the vendor and model IDs to save you from flashing incompatible FW

These firmware validation tests are really designed for millions of people as a sanity check, rather than a check for the ODMs to properly validate an image.

**Comment by YehezkelShB** [comment]
> > So you can't trust the vendor and model IDs to save you from flashing incompatible FW
> 
> These firmware validation tests are really designed for millions of people as a sanity check, rather than a check for the ODMs to properly validate an image.

OK, I understand your point. It makes sense when you look at it from this perspective.

BTW, sorry again for not having the time right now for deep diving into this review.

File: plugins/thunderbolt/fu-thunderbolt-firmware.c (hunk 9 above)
Comment (superm1) [reply plugins/thunderbolt/fu-thunderbolt-firmware.c:487]: I would prefer lets leave it, I think that this type of thing is potentially useful for pre-production hardware for integrators and developers to catch a problem.

Comment (superm1) [reply plugins/thunderbolt/fu-thunderbolt-firmware.c:487]: (It of course should be added to the output when parsing to make it useful though)

Comment (hughsie) [reply plugins/thunderbolt/fu-thunderbolt-firmware.c:487]: I've wired it in properly. @YehezkelShB good catch about it only being AR and TR.

Comment (YehezkelShB) [reply plugins/thunderbolt/fu-thunderbolt-firmware.c:525]: Still missing the "it was checked only for host" part of my comment. The current code does the opposite

**Comment by superm1** [comment]
@hughsie I think we'll probably want to slip this to later than 1.4.0.  Hopefully @gicmo can help with the testing framework changes and eventualy I'll work on updating the force power plugin to work with this.

My main challenge with force power is with WFH I don't have anything that uses force power anymore to test with.

**Comment by hughsie** [comment]
> I think we'll probably want to slip this to later than 1.4.0

Agree.

File: libfwupdplugin/fu-udev-device.c [hunk 10]
```diff
@@ -943,6 +943,83 @@ fu_udev_device_pwrite (FuUdevDevice *self, goffset port, guint8 data, GError **e
...
 
+/**
+ * fu_udev_device_get_parent_name
+ * @self: A #FuUdevDevice
+ *
+ * Returns the name of the direct ancestor of this device
+ *
+ * Returns: string or NULL if unset or invalid
+ *
+ * Since: 1.4.0
```
Comment (superm1) [inline libfwupdplugin/fu-udev-device.c:954]: need to fix this to 1.4.2

File: libfwupdplugin/fu-udev-device.c [hunk 11]
```diff
@@ -943,6 +943,83 @@ fu_udev_device_pwrite (FuUdevDevice *self, goffset port, guint8 data, GError **e
...
+ * fu_udev_device_get_sysfs_attr:
+ * @self: A #FuUdevDevice
+ * @attr: offset address
+ * @error: A #GError, or %NULL
+ *
+ * Reads an arbitrary sysfs attribute 'attr' associated with UDEV device
+ *
+ * Returns: string or NULL
+ *
+ * Since: 1.4.0
```
Comment (superm1) [inline libfwupdplugin/fu-udev-device.c:982]: need to fix this to 1.4.2

File: libfwupdplugin/fwupdplugin.map [hunk 12]
```diff
@@ -569,6 +569,8 @@ LIBFWUPDPLUGIN_1.4.0 {
     fu_hid_device_set_report;
     fu_plugin_get_config_value_boolean;
     fu_plugin_runner_device_created;
+    fu_udev_device_get_parent_name;
+    fu_udev_device_get_sysfs_attr;
```
Comment (superm1) [inline libfwupdplugin/fwupdplugin.map:573]: need to fix this to 1.4.2

**Comment by superm1** [comment]
@hughsie @gicmo so hypothetically I've got the test suite fixed up now to work with the plugin overhaul, can you review that part and make sure it's looking good?

The last thing left is fixing up force power to work again, which will have to be best effort since I don't have hardware using it anymore.

**Comment by superm1** [comment]
Btw the most difficult part to solve was understanding why change events were flowing through.  Turns out that gudev caches sysfs attributes!  I was going crazy trying to follow in a debugger why they weren't refreshing until I just straced the binary and noticed nothing called out to `nvm_version` again.
https://github.com/lu-zero/udev/blob/28f04cff8a5b4e4dd979438b7e91732a309400e7/src/libudev-device.c#L1286

File: libfwupdplugin/fu-udev-device.c [hunk 13]
```diff
@@ -943,6 +943,57 @@ fu_udev_device_pwrite (FuUdevDevice *self, goffset port, guint8 data, GError **e
 #endif
 }
 
+/**
+ * fu_udev_device_get_sysfs_attr:
+ * @self: A #FuUdevDevice
+ * @attr: offset address
```
Comment (hughsie) [inline libfwupdplugin/fu-udev-device.c:975]: is this true?

File: libfwupdplugin/fu-udev-device.c [hunk 14]
```diff
@@ -943,6 +943,32 @@ fu_udev_device_pwrite (FuUdevDevice *self, goffset port, guint8 data, GError **e
...
+fu_udev_device_get_parent_name (FuUdevDevice *self)
+{
+#ifdef HAVE_GUDEV
+	FuUdevDevicePrivate *priv = GET_PRIVATE (self);
+	g_autoptr(GUdevDevice) parent = NULL;
+
+	g_return_val_if_fail (FU_IS_UDEV_DEVICE (self), NULL);
+
+	parent = g_udev_device_get_parent (priv->udev_device);
+	return parent == NULL ? NULL :g_udev_device_get_name (parent);
```
Comment (hughsie) [inline libfwupdplugin/fu-udev-device.c:966]: The lifecycle here looks fishy. The `transfer:none` name of the parent belongs to the parent, but we're unreffing it when this function returns. I think this needs to be `g_strdup`'d and a `gchar *` returned.

File: libfwupdplugin/fu-plugin.c [hunk 15]
```diff
@@ -1694,6 +1694,19 @@ fu_plugin_usb_device_added (FuPlugin *self, FuUsbDevice *device, GError **error)
...
+fu_plugin_udev_device_changed (FuPlugin *self, FuUdevDevice *device, GError **error)
+{
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	/* open */
+	locker = fu_device_locker_new (FU_DEVICE (device), error);
+	if (locker == NULL)
+		return FALSE;
+	return fu_device_rescan (FU_DEVICE (device), error);
+
```
Comment (hughsie) [inline libfwupdplugin/fu-plugin.c:1707]: random blank line

File: libfwupdplugin/fu-plugin.c [hunk 16]
```diff
@@ -1872,8 +1885,14 @@ fu_plugin_runner_udev_device_changed (FuPlugin *self, FuUdevDevice *device, GErr
 
 	/* optional */
 	g_module_symbol (priv->module, "fu_plugin_udev_device_changed", (gpointer *) &func);
-	if (func == NULL)
+	if (func == NULL) {
+		if (priv->device_gtype != G_TYPE_INVALID ||
+		    fu_device_get_specialized_gtype (FU_DEVICE (device)) != G_TYPE_INVALID) {
```
Comment (hughsie) [inline libfwupdplugin/fu-plugin.c:1890]: do we need both? I was surprised to see the device_gtype one.

**Comment by hughsie** [comment]
> The last thing left is fixing up force power to work again

I guess my only concern is if there are any vendors shipping TBT updates on the LVFS that require force power. Obviously you know for Dell, but what about Lenovo for example? What's the end-user-visible effect of trying to deploy a tbt update on a tbt controller that needs force power? If I understand correctly the tbt controller will just be "missing" without the force power stuff, unless a tbt device is attached. If that's the only "regression" I think that's acceptable.

File: libfwupdplugin/fu-plugin.c (hunk 16 above)
Comment (superm1) [reply libfwupdplugin/fu-plugin.c:1890]: I was modeling off of the usb ones, which yes do.

**Comment by superm1** [comment]
> > The last thing left is fixing up force power to work again
> 
> I guess my only concern is if there are any vendors shipping TBT updates on the LVFS that require force power. Obviously you know for Dell, but what about Lenovo for example? What's the end-user-visible effect of trying to deploy a tbt update on a tbt controller that needs force power? If I understand correctly the tbt controller will just be "missing" without the force power stuff, unless a tbt device is attached. If that's the only "regression" I think that's acceptable.

That's correct, the obvious regression is that you can't update the device unless it's powered by something being plugged in.  I was also thinking this is an acceptable regression.

I'm not aware of any other impact though.

File: libfwupdplugin/fu-udev-device.c (hunk 14 above)
Comment (superm1) [reply libfwupdplugin/fu-udev-device.c]: fixed

File: libfwupdplugin/fu-udev-device.c (hunk 13 above)
Comment (superm1) [reply libfwupdplugin/fu-udev-device.c]: fixed

File: plugins/thunderbolt/fu-thunderbolt-firmware.c (hunk 9 above)
Comment (superm1) [reply plugins/thunderbolt/fu-thunderbolt-firmware.c]: I've fixed this in the current commit and reversed it back to only checking if on a host controller, thanks for catching this.

File: plugins/thunderbolt/fu-thunderbolt-firmware.c (hunk 7 above)
Comment (superm1) [reply plugins/thunderbolt/fu-thunderbolt-firmware.c]: I've adjusted the comment to make it clear that we shouldn't actually use the value here but just match it between images.

File: plugins/thunderbolt/fu-thunderbolt-device.c (hunk 2 above)
Comment (superm1) [reply plugins/thunderbolt/fu-thunderbolt-device.c:180]: Added these.

**Comment by superm1** [comment]
Thanks, will merge and continue to test with it.
